design/version-vector.md

```diff
@@ -0,0 +1,66 @@
+
+### Goal
+
+Introduce yorkie's logical clock system.
+
+### Non Goals
+
+This document will not explain how CRDT works and Yorkie's document structure.
+
+## Proposal Details
+
+### What is Lamport Timestamp
+`Lamport Timestamps` are a logical clock mechanism used in distributed systems to order events without relying on synchronized physical clocks. Each process in the system maintains a counter that increments with every event, and when a process sends a message, it includes its current counter value. Upon receiving a message, the recipient updates its own counter to the maximum of its current value or the received timestamp, ensuring a consistent ordering of events. `Lamport Timestamps` provide a way to determine the "happened-before" relationship between events, helping in coordinating actions in a distributed environment.
+```
+Here's a simplified example of Lamport Timestamps in a distributed system with three processes (P1, P2, P3):
+
+1. P1 performs an event: P1's clock becomes 1.
+2. P1 sends a message to P2: P1 increments its clock to 2 and sends the message with timestamp 2.
+3. P2 receives the message: P2 updates its clock to max(0, 2) + 1 = 3.
+4. P2 sends a message to P3: P2's clock becomes 4.
+5. P3 receives the message: P3 updates its clock to max(0, 4) + 1 = 5.
+6. P3 sends a message to P1: P3's clock becomes 6.
+7. P1 receives the message: P1 updates its clock to max(2, 6) + 1 = 7.
+8. The timestamps help order events across processes even without synchronized physical clocks.
+```
+
+### Brief Explain of Yorkie's Logical Clock System
+All updates are recorded as `change`, and the ID of the `change` is composed of a `Lamport Timestamp` and `actorID` called `TimeTicket`. The ID of each CRDT node created through `changes` is also composed of a `TimeTicket`.
+
+### The reason for choosing Version Vector To Replace Lamport Based Logical Clock System
+Yorkie provides Garbage Collection for removed (tombstone) nodes. The previously implemented Lamport-based ID system triggers a crucial problem with causality.
+You can see detail from [garbage collection design document](./garbage-collection.md).
+
+### What is Version Vector
+
+A Version Vector is a mechanism used in distributed systems to track the causal relationships between different replicas or nodes. Each node maintains a vector, where each element corresponds to a version counter for every other node in the system, including itself. Whenever a node updates its state, it increments its own counter. When nodes exchange information, they merge their vectors by taking the maximum value for each element, ensuring that the system knows which updates happened concurrently or which one came first. Version Vectors are particularly useful for detecting conflicts and maintaining consistency in distributed databases by providing a more detailed view of the system's state than a single logical clock.
+
+![version-vector-example](media/version-vector-example.jpg)
```

Comment:
```
Whenever a node updates its state, it increments its own counter. When nodes exchange information, they merge their vectors by taking the maximum value for each element, 
```
This part can be confusing because the case that you've mentioned is `Vector Clock`.
Also, implementation can be diverged but the important part is maintaining the concept of Version Vector. 
version at B@3 is always identical at any time. 

For your information, here is a specific explanation. A's lamport increases at A@2, and when B gets A@2, it'll be max([2, 0, 0], [0, 0, 0])  => [2, 0, 0], then we need to compute B's new lamport which is max(2, 1).

Reply:
thx! understand it!